clinicadl/generate/generate_config.py

```diff
@@ -0,0 +1,175 @@
+from enum import Enum
+from logging import getLogger
+from pathlib import Path
+from typing import Annotated, Union
+
+from pydantic import BaseModel, field_validator
+
+from clinicadl.utils.exceptions import ClinicaDLTSVError
+
+logger = getLogger("clinicadl.predict_config")
+
+
+class Preprocessing(str, Enum):
+    """Possible preprocessing method in clinicaDL."""
+
+    T1_LINEAR = "t1-linear"
+    T1_EXTENSIVE = "t1-extensive"
+    PET_LINEAR = "pet-linear"
+
+
+class SUVRReferenceRegions(str, Enum):
+    """Possible SUVR reference region for pet images in clinicaDL."""
+
+    PONS = "pons"
+    CEREBELLUMPONS = "cerebellumPons"
+    PONS2 = "pons2"
+    CEREBELLUMPONS2 = "cerebellumPons2"
+
+
+class Tracer(str, Enum):
+    """Possible tracer for pet images in clinicaDL."""
+
+    FFDG = "18FFDG"
+    FAV45 = "18FAV45"
+
+
+class Pathology(str, Enum):
+    """Possible pathology for hypometabolic generation of pet images in clinicaDL."""
+
+    AD = "ad"
+    BVFTD = "bvftd"
+    LVPPA = "lvppa"
+    NFVPPA = "nfvppa"
+    PCA = "pca"
+    SVPPA = "svppa"
+
+
+class GenerateConfig(BaseModel):
+    generated_caps_directory: Path = Path("")
+    n_subjects: int = 300
+    n_proc: int = 1
+
+
+class SharedGenerateConfigOne(GenerateConfig):
+    caps_directory: Path = Path("")
+    participants_list: Path = Path("")
+    preprocessing_cls: Preprocessing = Preprocessing.T1_LINEAR
+    use_uncropped_image: bool = False
+
+    @field_validator("participants_list", mode="before")
+    def check_tsv_file(cls, v):
+        if not isinstance(v, Path):
+            Path(v)
+        # if not v.is_file():
+        #     raise ClinicaDLTSVError(
+        #         "The participants_list you gave is not a file. Please give an existing file."
+        #     )
+        # if v.stat().st_size == 0:
+        #     raise ClinicaDLTSVError(
+        #         "The participants_list you gave is empty. Please give a non-empty file."
+        #     )
+
+        return v
+
+    @property
+    def preprocessing(self) -> Preprocessing:
+        return self.preprocessing_cls.value
+
+    @preprocessing.setter
+    def preprocessing(self, value: Union[str, Preprocessing]):
+        self.preprocessing_cls = Preprocessing(value)
+
+
+class SharedGenerateConfigTwo(SharedGenerateConfigOne):
+    suvr_reference_region_cls: SUVRReferenceRegions = SUVRReferenceRegions.PONS
+    tracer_cls: Tracer = Tracer.FFDG
+
+    @property
+    def suvr_reference_region(self) -> SUVRReferenceRegions:
+        return self.suvr_reference_region_cls.value
```

Comment:
Unless I missed something, this should be:

```suggestion
        return self.suvr_reference_region_cls
```

Otherwise, it returns the string value of the enumeration variant.

Also, you could spare the `cls` suffix and have simply a private attribute `_suvr_reference_region` which can only be read and modified through the getters and setters which handle the validation.
Does this make sense ?

This would apply to tracer as well.

Reply:
Hi @thibaultdvx 

Sure, I can explain a bit what I meant.
I would recommend having all functions expecting or returning SUVR regions to use the enumeration because it is much safer than a plain string and would avoid having validation logic all over the code base.

Nonetheless, when a function (or class) belongs to the public API of the package (i.e. users of the package are allowed and assumed to use it in their own code), it might be convenient to have a union type `str | SUVRReferenceRegions` such that they can pass simple strings instead of having to import the enum themselves.

```python
def user_facing_func(region: str | SUVRReferenceRegions):
    # Perform the input validation (will raise if an invalid string was passed)
    region = SUVRReferenceRegions(region)
    # from now on, there is no validation to do, we know that
    # region is a valid SUVR region value
    _do_stuff(region)
    ...

def _do_stuff(region: SUVRReferenceRegions): # no string, only a SUVRReferenceRegions type here
    # no validation to do here
    if region == SUVRReferenceRegions.VALUE1:
        _do_other_stuff()
    if region == SUVRReferenceRegions.VALUE2:
        _do_other_stuff_2()
    # no need to raise a value error here, region MUST be one of its variant...
```

Let me know if this is not clear.

Reply:
You'd use the `value` attribute when needing to access the string value of the enumeration variant. For example, when printing, you probably want to print the string value and not the string representation of the enumeration variant:

```python
def print_suvr(region: SUVRReferenceRegions):
    # this is probably what you want:
    print(f"The region is {region.value}")
    # This is probably NOT what you want:
    print(f"The region is {region}")
```

Reply:
@NicolasGensollen thanks a lot. It is perfectly clear and I think this is what we want for ClinicaDL.